### ISSUE TITLE
API Mock-up

### ISSUE THREAD

**Issue by premun** [sharpliner/sharpliner#11]
This is a living issue. Edit the description directly.

```csharp
public class MainPipeline : Sharpliner.PipelineDefinition
{
    protected override string TargetFile => "/azure-pipelines.yml";

    protected override Pipeline Pipeline => new()
    {
        Name = "$(Date:yyyMMdd).$(Rev:rr)",

        Triggers = new[]
        {
            new()
            {
                Branches = new[] { "main", "production" },
                IsBatch = true,
                Exclude = new[] { "tools/*" },
            }
        },

        Prs = new[]
        {
            new()
            {
                Branches = new[] { "main" },
                AutoCancel = true,
            }
        },

        Variables = new[]
        {
            Variable("Configuration", "Release"),
            Variable("DOTNET_SKIP_FIRST_TIME_EXPERIENCE", true),
            Group("PR keyvault variables"),

            If.Equal("variables['Build.Reason']", "PullRequest")
                .Variable("TargetBranch", "$(System.PullRequest.SourceBranch)"),

            If.NotEqual("variables['Build.Reason']", "PullRequest")
                .Variable("TargetBranch", "$(Build.SourceBranch)"),

            If(And(Equals("variables['Build.SourceBranch']", "refs/heads/production"), NotEquals("Configuration", "Debug")))
                .Variable("PublishProfileFile", "Prod")
                .Variable("AzureSubscription", "NetHelix"),
        },

        Stages = new[]
        {
            new()
            {
                Jobs = new[]
                {
                    new Job("PublishClient", "Publish wheel packages")
                    {
                        Timeout = TimeSpan.FromMinutes(30),
                        Pool = new HostedPool(vmImage: "windows-2019"),
                        Checkout = 'self',
                        Clean = true,
                        Steps = new[]
                        {
                            new BashTask("Install wheel and twine"
                                "pip install wheel==0.33.6 twine==2.0.0"),

                            new TwineAuthenticateTask(artifactFeed: "public/$(HelixClientFeedName)"),

                            new BashTask("Publish Helix Installer Wheel"
                                @"python $(HelixInstallerDir)\setup.py egg_info --tag-build $(HelixScriptsVersion) bdist_wheel
                                  python -m twine upload -r $(HelixClientFeedName) --config-file $(PYPIRC_PATH) $(HelixInstallerDir)\dist\*.whl"),
                        ]
                    },

                    new Job("BuildClient", "Build solution")
                    {
                        Timeout = TimeSpan.FromMinutes(30),
                        Pool = new HostedPool(vmImage: "windows-2019"),
                        Checkout = 'self',
                        Clean = true,
                        DependsOn = new[] { "PublishClient" },
                        Steps = new[]
                        {
                            If("and(eq(variables['Build.Reason'], 'PullRequest'), ne(variables['Build.SourceBranch'], 'refs/heads/production'))",
                                new PowerShellTask("Enforce GitHub issue link presence"
                                    "eng/enforce-issue.ps1 -PullRequestId $(System.PullRequest.PullRequestId) -AccessToken $(System.AccessToken)")),

                            new UseDotNetTask("Use .NET 2.1.x SDK", type: DotNetPackageType.SDK, version: "2.1.x"),

                            new CustomDotNetCoreCLITask("Build Helix.Machines.sln", "msbuild") {
                                Projects = new[] { "Helix.Machines.sln" },
                                Arguments = new[] { "-v:m", "-m:1", "-nr:false", "-restore", "-t:Build;Publish" },
                            }
                        }
                    }
                }
            }
        }
    };
}
```

This is the source pipeline that we are trying to model:
```yaml
name: $(Date:yyyMMdd).$(Rev:rr)

trigger:
  batch: true
  branches:
    include:
      - main
      - production
  paths:
    exclude:
      - tools/*

pr:
  autoCancel: true
  branches:
    include:
      - main
  paths:
    exclude:
      - tools/*

variables:
  - name: Configuration
    value: Release
  - name: DOTNET_SKIP_FIRST_TIME_EXPERIENCE
    value: true
  - group: PR keyvault variables

  - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
      - name: TargetBranch
        value: $(System.PullRequest.SourceBranch)
  - ${{ if ne(variables['Build.Reason'], 'PullRequest') }}:
      - name: TargetBranch
        value: $(Build.SourceBranch)

  - ${{ if ne(variables['Build.SourceBranch'], 'refs/heads/production') }}:
      - name: PublishProfileFile
        value: Prod
      - name: AzureSubscription
        value: NetHelix

stages:
  - stage: Build
    jobs:
      - job: PublishClient
        displayName: Publish wheel packages
        timeoutInMinutes: 30
        pool:
          vmImage: "windows-2019"

        steps:
          - script: pip install wheel==0.33.6 twine==2.0.0
            displayName: Install wheel and twine

          - task: TwineAuthenticate@1
            displayName: Twine Authenticate
            inputs:
              artifactFeed: public/$(HelixClientFeedName)

          - script: |
              python $(HelixInstallerDir)\setup.py egg_info --tag-build $(HelixScriptsVersion) bdist_wheel
              python -m twine upload -r $(HelixClientFeedName) --config-file $(PYPIRC_PATH) $(HelixInstallerDir)\dist\*.whl
            displayName: Create/Publish Helix Installer Wheel

      - job: Build
        displayName: Build solution
        dependsOn: PublishClient
        timeoutInMinutes: 30
        pool:
          vmImage: "windows-2019"

        steps:
          - checkout: self
            clean: true

          - ${{ if and(eq(variables['Build.Reason'], 'PullRequest'), ne(variables['Build.SourceBranch'], 'refs/heads/production')) }}:
              - powershell: eng/enforce-issue.ps1 -PullRequestId $(System.PullRequest.PullRequestId) -AccessToken $(System.AccessToken)
                displayName: Enforce GitHub issue link presence

          - task: UseDotNet@2
            displayName: Use .NET 2.1.x SDK
            inputs:
              packageType: sdk
              version: 2.1.x

          - task: DotNetCoreCLI@2
            inputs:
              command: custom
              custom: msbuild
              projects: Helix.Machines.sln
              arguments: >-
                -v:m
                -m:1
                -nr:false
                -restore
                -t:Build;Publish
```

**Comment by mandel-macaque** [comment]
Comments about the above:

### Triggers

I would go the  declarative way form
```csharp
Trigger = Trigger.ForBranch("main", "production").Batch().Exclude("tools/*"),
```
to 
```csharp
Trigger = {
    new {
        Branch = {"main", "production"}
        IsBatch = true,
       Exclude = { "tools/*" }
   },
}
```
Also, should be a collection, we can have more than one trigger AFAIK

### PRs

Same comment as with triggers.

**Comment by premun** [comment]
Nice! Everything makes sense:
- No more mixing fluent and declaration
- Id validation can happen for strings too
- Strong-typed logic operators are also nice!

**Comment by premun** [comment]
Also, the definition itself is now shorter in C# than in YAML. Even with all those closing and opening brackets.

**Comment by mandel-macaque** [comment]
Yes! We need to find the right balance between the two approaches. I think that adjectives should be all fluent, while properties should all be declarative even if optional. Then use declarative for all conditions. The question is what reads better:
```csharp
If().Equals("varname", "value")
    .Variable("conditionalVar","value");
```
Or
```csharp
If(Equals("varname", "value"))
    .Variable("condition","value");
```
I'm going for the second one, because with complex expressions it looks shorter/nicer:
```csharp
If(And(Equals("var", "value"),Equals("var","value")))
    .Variable("var","value");
```
We should look at Nunit and xUnit for ideas since they have a lot of assertions and more experience. I think with some smart use of genetics this will me very doable.

PS: I think will be shorter because the concatenation of several ands would be very verbose

**Comment by mandel-macaque** [comment]
Also, thinking about serialization, 2 is closer to the one used by yaml

**Comment by premun** [comment]
👍 Makes sense with adjectives
👍 Second one looks almost like YAML so yeah, let's do that one

**Comment by premun** [comment]
@mandel-macaque another idea:

We could move the top level properties of the Pipeline POCO to the `PipelineDefinition` class itself and the user would override those. Rather than having:
```csharp
    protected override Pipeline Pipeline => new()
    {
        Name = "$(Date:yyyMMdd).$(Rev:rr)",

        Variables = new[]
        {
            Variable("Configuration", "Release"),
        },
    }
```
you would do
```csharp
    protected override string Name => "$(Date:yyyMMdd).$(Rev:rr)";

    protected override IEnumerable<Variable> Variables => new[]
    {
        Variable("Configuration", "Release"),
    };
```

But I guess that will be more confusing?

**Comment by mandel-macaque** [comment]
I'd go for less typing, also, we can let people define several pipelines if needed and reference them as resources (yes, is a thing and very useful).

**Comment by premun** [comment]
I have the API for variable definition ready but I will be moving rooms this weekend so I might be offline.

If you want to have a look or work on top of that it's in the api branch in my fork:
https://github.com/premun/sharpliner/blob/prvysoky/api/tests/Sharpliner.Serialization.Tests/MockPipeline.cs

**Comment by premun** [comment]
I have this for nested ifs:

![image](https://user-images.githubusercontent.com/7013027/115106732-a80f7d00-9f66-11eb-9e9a-fcd60c637b2c.png)

**Comment by premun** [comment]
Closing as it was implemented already in #14 and #15
